### PR TITLE
First pass at a public composite pattern

### PR DESCRIPTION
**Pull request by jskeet**
This is designed to provide a way of fixing #607.
I'm not entirely convinced about the design though... the "parse
using patterns in one direction, format using patterns in the other
direction" is useful for us internally (because we need partial
patterns), but not obvious.

Putting out there for comment... I'll add more tests once we've got
the design sorted.

### CODE REVIEW

**Comment by jskeet** [comment]
cc @lundmikkel and @malcolmr - this PR isn't ready for submission, but is ready for design consideration.

File: src/NodaTime/Utility/Preconditions.cs [hunk 1]
```diff
@@ -112,11 +112,13 @@ internal static void DebugCheckArgumentRange([InvokerParameterName] string param
         [Conditional("DEBUG")]
         internal static void DebugCheckArgument(bool expression, [InvokerParameterName] string parameter, string messageFormat, params object[] messageArgs)
         {
+#if DEBUG
```
Comment (malcolmr) [inline src/NodaTime/Utility/Preconditions.cs]: Changes in this file seem useful but unrelated?

Comment (jskeet) [inline src/NodaTime/Utility/Preconditions.cs]: I needed to introduce DebugCheckState, and that led me to find the inconsistent exception use. It may well be worth me pulling out these changes into a separate PR though...

File: src/NodaTime/Text/CompositePatternBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    /// A composite pattern is a combination of multiple patterns. When parsing, these are checked
+    /// in the order in which they are added to the builder with the <see cref="Add(IPattern{T}, Func{T, bool})"/>
+    /// method, by trying to parse and seeing if the result is a successful one. When formatting,
+    /// the patterns are checked in the reverse order, using the predicate provided along with the pattern
+    /// when calling <c>Add</c>. The intention is that patterns are added in "most precise first" order,
+    /// and the predicate should indicate whether it can fully represent the given value - so the "less precise"
+    /// (and therefore usually shorter) pattern can be used first.
+    /// </remarks>
+    /// <typeparam name="T">The type of value to be parsed or formatted by the resulting pattern.</typeparam>
+    internal sealed class CompositePatternBuilder<T> : IEnumerable<IPattern<T>>
```
Comment (jskeet) [inline src/NodaTime/Text/CompositePatternBuilder.cs]: This was meant to be public... but if we submit while it's internal, we can still reflect on the API a bit further.

I'll add some more tests, fix whitespace-only changes and then merge.
